File: framework/languages/java/samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/adapters/zlink/spots/handlers/MatchBingoActorHandler.java
package systems.zlink.samples.bingo.server.play.adapters.zlink.spots.handlers;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.CancellationToken;
import systems.zlink.framework.actors.ZLinkActorJoinResult;
import systems.zlink.framework.spots.ZLinkEntrySpotActorRequestHandler;
import systems.zlink.framework.spots.ZLinkSpotActorRequestContext;
import systems.zlink.samples.bingo.server.play.adapters.zlink.actors.PlayerActor;
import systems.zlink.samples.bingo.server.play.adapters.zlink.spots.BingoEntrySpot;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.server.configuration.SampleTopology;
import systems.zlink.samples.bingo.shared.contracts.Messages;

public final class MatchBingoActorHandler
    implements ZLinkEntrySpotActorRequestHandler<
        BingoEntrySpot,
        PlayerActor,
        Messages.MatchBingoReq,
        Messages.MatchBingoRes> {
    @Override
    public Messages.MatchBingoRes handle(
        BingoEntrySpot entrySpot,
        PlayerActor actor,
        ZLinkSpotActorRequestContext context,
        Messages.MatchBingoReq request,
        CancellationToken cancellationToken) {
        Messages.MatchBingoApiRes matched;
        try {
            System.out.println("bingo match: api request actor=" + actor.actorId());
            matched = entrySpot.context().outbound().requestToChannel(
                    SampleNames.ApiChannel,
                    new Messages.MatchBingoApiReq(
                        actor.actorId(),
                        actor.displayName(),
                        request.mode(),
                        SampleTopology.selectedPlayNodeRid()))
                .timeout(SampleTimings.RequestTimeout)
                .await(Messages.MatchBingoApiRes.class);
            System.out.println("bingo match: allocated room=" + matched.roomId());
        } catch (RuntimeException ex) {
            ex.printStackTrace(System.out);
            throw ex;
        }

        if (cancellationToken.isCancellationRequested()) {
            throw new IllegalStateException("MatchBingoReq was cancelled");
        }
        ZLinkActorJoinResult<Messages.BingoRoomJoinRes> joined;
        try {
            System.out.println("bingo match: join room=" + matched.roomId() + " actor=" + actor.actorId());
            joined = actor.context()
                .joinSpot(
                    RoutingId.from(matched.roomId()),
                    new Messages.BingoRoomJoinReq(
                        matched.roomId(),
                        actor.actorId(),
                        actor.displayName(),
                        false))
                .timeout(SampleTimings.RequestTimeout)
                .await(Messages.BingoRoomJoinRes.class);
            System.out.println("bingo match: joined room=" + matched.roomId() + " actor=" + actor.actorId());
        } catch (RuntimeException ex) {
            ex.printStackTrace(System.out);
            throw ex;
        }
        if (cancellationToken.isCancellationRequested()) {
            throw new IllegalStateException("MatchBingoReq was cancelled");
        }
        return new Messages.MatchBingoRes(
            matched.roomId(),
            joined.reply().state(),
            matched.roomOwnerNodeRid());
    }
}

File: framework/languages/java/samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/adapters/zlink/spots/handlers/MatchBingoActorHandler.kt
package systems.zlink.samples.kotlin.bingo.server.play.adapters.zlink.spots.handlers

import kotlinx.coroutines.future.await
import systems.zlink.contracts.core.RoutingId
import systems.zlink.framework.CancellationToken
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.framework.spots.ZLinkSpotActorRequestContext
import systems.zlink.samples.kotlin.bingo.server.play.adapters.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.server.play.adapters.zlink.spots.BingoEntrySpot
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTopology
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameStartedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoRes

class MatchBingoActorHandler() : ZLinkSuspendingEntrySpotActorRequestHandler<
    BingoEntrySpot,
    PlayerActor,
    MatchBingoReq,
    MatchBingoRes,
    > {
    override suspend fun handle(
        entrySpot: BingoEntrySpot,
        actor: PlayerActor,
        context: ZLinkSpotActorRequestContext,
        request: MatchBingoReq,
        cancellationToken: CancellationToken,
    ): MatchBingoRes {
        val matched = entrySpot.context().outbound().requestToChannel(
            SampleNames.ApiChannel,
            MatchBingoApiReq(
                actor.actorId(),
                actor.displayName,
                request.mode,
                SampleTopology.selectedPlayNodeRid(),
            ),
        )
            .timeout(SampleTimings.RequestTimeout)
            .submit(MatchBingoApiRes::class.java)
            .await()
        if (cancellationToken.isCancellationRequested) {
            throw IllegalStateException("MatchBingoReq was cancelled")
        }
        val joined = actor.context()
            .joinSpot(
                RoutingId.from(matched.roomId),
                BingoRoomJoinReq(
                    matched.roomId,
                    actor.actorId(),
                    actor.displayName,
                    false,
                ),
            )
            .timeout(SampleTimings.RequestTimeout)
            .submit(BingoRoomJoinRes::class.java)
            .await()
        if (cancellationToken.isCancellationRequested) {
            throw IllegalStateException("MatchBingoReq was cancelled")
        }
        if (joined.reply().state.status == "Running") {
            actor.context()
                .boundSession()
                .send(BingoGameStartedNotify(joined.reply().state))
                .submit()
                .await()
        }
        return MatchBingoRes(matched.roomId, joined.reply().state, matched.roomOwnerNodeRid)
    }
}
